Release v5.7.0 (#6615)
OpenZeppelin/openzeppelin-contracts: CHANGELOG.md

#### Access

- `AccessManager`: Treat `setAuthority` differently in `canCall` to prevent bypassing the `updateAuthority` security using an `execute`. ([#6388](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6388))

> - `AccessManager`: Allow a role admin to cancel grant and revoke operations. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

Reviewer: ```suggestion
- `AccessManager`: Allow a role admin to cancel, grant, and revoke operations. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))
```

> - `AccessManager`: Allow a role admin to cancel grant and revoke operations. ([#6573](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6573))

Reviewer 2: ```suggestion
- `AccessManager`: Allow a role admin to cancel both grant and revoke operations. ([#6573](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6573))
```

> - `AccessManager`: Allow a role admin to cancel grant and revoke operations. ([#6573](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6573))

Reviewer: ```suggestion
- `AccessManager`: Allow a role admin to cancel operations that grant or revoke roles. ([#6573](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6573))
```

> - `AccessManager`: Allow a role admin to cancel, grant, and revoke operations. ([#6573](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6573))

Reviewer 2: ```suggestion
- `AccessManager`: Allow a role admin to cancel grant and revoke operations. ([#6573](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6573))
```

#### Account

> - `AccountERC7579`: Revert the uninstallation of any module (validator, executor, fallback, or hook) if its `onUninstall` callback reverts, giving modules control over their own uninstallation. A forced uninstallation that bypasses the callback can still be performed through a delegate call via `execute`. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

Reviewer 3: ```suggestion
- `AccountERC7579`: Revert the uninstallation of any module (validator, executor, fallback, or hook) if its `onUninstall` callback reverts, giving modules control over their own uninstallation. A forced uninstallation that bypasses the callback can still be performed through a delegate call via `execute`. ([#6628](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6628))
```

#### Cryptography

> - `EIP712`: Drop the storage fallback for long `name`/`version` values. Both parameters must now fit in a `ShortString` (at most 31 bytes) or the constructor reverts with `ShortStrings.StringTooLong`. Storing the domain exclusively in immutables keeps the domain (and downstream `ERC7739` verification) consistent when the contract is used behind a proxy or clone without an initializer. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

Reviewer 3: ```suggestion
- `EIP712`: Drop the storage fallback for long `name`/`version` values. Both parameters must now fit in a `ShortString` (at most 31 bytes) or the constructor reverts with `ShortStrings.StringTooLong`. Storing the domain exclusively in immutables keeps the domain (and downstream `ERC7739` verification) consistent when the contract is used behind a proxy or clone without an initializer. ([#6631](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6631))
```

- `ERC7739`: Reject signatures whose `contentsDescr` fails to parse into a non-empty `contentsName`, preventing a malformed descriptor from degrading verification to a constant `structHash` that no longer binds the message contents or the account's EIP-712 domain. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))
- `ERC7913WebAuthnVerifier`: Add an internal `_requireUV` function that can be overridden to disable the UV check. ([#6596](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6596))

> - `MultiSignerERC7913`: Decode the multisignature payload directly from calldata and return `false` on malformed encoding instead of reverting during `abi.decode`, so paymaster/account validation can surface `SIG_VALIDATION_FAILED` rather than bubble up a revert. The `_validateSignatures` and `_validateThreshold` override parameters change from `bytes[] memory` to `bytes[] calldata`. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

Reviewer 3: ```suggestion
- `MultiSignerERC7913`: Decode the multisignature payload directly from calldata and return `false` on malformed encoding instead of reverting during `abi.decode`, so paymaster/account validation can surface `SIG_VALIDATION_FAILED` rather than bubble up a revert. The `_validateSignatures` and `_validateThreshold` override parameters change from `bytes[] memory` to `bytes[] calldata`. ([#6642](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6642))
```

- `RSA`: Return `false` from `pkcs1Sha256` instead of reverting when modular exponentiation fails. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

> - `SignatureChecker`: Zero-pad the ERC-1271 signature calldata to a 32-byte boundary when performing the ERC-1271 static call, so the encoded `bytes` argument conforms to the ABI spec. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

Reviewer 3: ```suggestion
- `SignatureChecker`: Zero-pad the ERC-1271 signature calldata to a 32-byte boundary when performing the ERC-1271 static call, so the encoded `bytes` argument conforms to the ABI spec. ([#6646](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6646))
```

- `BridgeMultiToken` and `BridgeERC1155`: Add bridge contracts to handle crosschain movements of ERC-1155 tokens. ([#6281](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6281))
- `CrosschainRemoteExecutor`: Add a new executor contract that relays transaction from a controller on a remote chain. ([#6272](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6272))

> #### Governance 

Reviewer 4: ```suggestion
#### Governance
```

- `BridgeMultiToken` and `BridgeERC1155`: Add bridge contracts to handle crosschain movements of ERC-1155 tokens. ([#6281](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6281))
- `CrosschainRemoteExecutor`: Add a new executor contract that relays transaction from a controller on a remote chain. ([#6272](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6272))

#### Governance

- `Governor`: Strictly enforce the expected proposal state depending on `proposalNeedsQueuing` when calling `execute`. ([#6386](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6386))
- `GovernorCrosschain`: Add governor module that facilitates the execution of crosschain operations through CrosschainRemoteExecutors and ERC-7786 gateways. ([#6272](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6272))

> - `GovernorPreventLateQuorum`: Bound `lateQuorumVoteExtension` by a new internal virtual `_maxLateQuorumVoteExtension` (default `votingPeriod()`) to cap the total voting duration to twice the voting period, thus preventing a large extension from bricking governance. Integrators can override `_maxLateQuorumVoteExtension` to enforce a different bound. ([#6615](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6615))

Reviewer 3: ```suggestion
- `GovernorPreventLateQuorum`: Bound `lateQuorumVoteExtension` by a new internal virtual `_maxLateQuorumVoteExtension` (default `votingPeriod()`) to cap the total voting duration to twice the voting period, thus preventing a large extension from bricking governance. Integrators can override `_maxLateQuorumVoteExtension` to enforce a different bound. ([#6644](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6644))
```
